File: examples/examples.ml
(* Parsing a JSON string using Js.Json.parse *)

let arrayOfInts str =
  let json = Js.Json.parseExn str in
  Json.Decode.(array int json)

(* prints `[3, 2, 1]` *)
let _ = Js.log (arrayOfInts "[1, 2, 3]" |> Js.Array.reverseInPlace)


(* Encoding a JSON data structure using Json.Encode *)

(* prints ["foo", "bar"] *)
let _ =
  [| "foo"; "bar" |]
  |> Json.Encode.stringArray
  |> Js.Json.stringify
  |> Js.log

(* prints ["foo", "bar"] *)
let _ =
  [| "foo"; "bar" |]
  |> Js.Array.map Json.Encode.string
  |> Json.Encode.array
  |> Js.Json.stringify
  |> Js.log


(* Decoding a fixed JSON data structure using Json.Decode *)
let mapJsonObjectString f decoder (encoder: int -> Js.Json.t) str =
  let json = Js.Json.parseExn str in
  Json.Decode.(dict decoder json)
    |> Js.Dict.map ((fun v -> f v) [@bs])
    |> Js.Dict.map ((fun v -> encoder v) [@bs])
    |> Json.Encode.object_
    |> Js.Json.stringify

let sum =
  Array.fold_left (+) 0

(* prints `{ "foo": 6, "bar": 24 }` *)
let _ =
  Js.log @@
    (mapJsonObjectString sum Json.Decode.(array int) Json.Encode.int {|
      {
        "foo": [1, 2, 3],
        "bar": [9, 8, 7]
      }
    |})


(* complex example *)

type address = {
  city: string;
  state: string
}

type person = {
  id: int;
  name: string;
  age: int option;
  address: address option
}

let parseAddress json: address =
  let open Json.Decode in {
    city = json |> field "city" string;
    state = json |> field "state" string
  }

let parsePerson json: person =
  let open Json.Decode in {
    id = json |> field "id" int;
    name = json |> field "name" string;
    age = json |> optional (field "age" int);
    address = json |> optional (field "address" parseAddress)
  }


let data = "{\"id\":1,\"name\":\"brad\",\"age\":27,\"address\":{\"city\":\"city1\",\"state\":\"state1\"}}"

let _ =
  Js.Json.parseExn data
  |> parsePerson
  |> Js.log

File: lib/js/examples/examples.js
'use strict';

var $$Array     = require("bs-platform/lib/js/array.js");
var Curry       = require("bs-platform/lib/js/curry.js");
var Js_dict     = require("bs-platform/lib/js/js_dict.js");
var Json_decode = require("../src/json_decode.js");

function arrayOfInts(str) {
  var json = JSON.parse(str);
  return Json_decode.array(Json_decode.$$int, json);
}

console.log(arrayOfInts("[1, 2, 3]").reverse());

console.log(JSON.stringify(/* array */[
          "foo",
          "bar"
        ]));

console.log(JSON.stringify(/* array */[
            "foo",
            "bar"
          ].map(function (prim) {
              return prim;
            })));

function mapJsonObjectString(f, decoder, encoder, str) {
  var json = JSON.parse(str);
  return JSON.stringify(Js_dict.map(Curry.__1(encoder), Js_dict.map(Curry.__1(f), Json_decode.dict(decoder, json))));
}

function sum(param) {
  return $$Array.fold_left(function (prim, prim$1) {
              return prim + prim$1 | 0;
            }, 0, param);
}

console.log(mapJsonObjectString(sum, function (param) {
          return Json_decode.array(Json_decode.$$int, param);
        }, function (prim) {
          return prim;
        }, "\n      {\n        \"foo\": [1, 2, 3],\n        \"bar\": [9, 8, 7]\n      }\n    "));

function parseAddress(json) {
  return /* record */[
          /* city */Json_decode.field("city", Json_decode.string, json),
          /* state */Json_decode.field("state", Json_decode.string, json)
        ];
}

function parsePerson(json) {
  return /* record */[
          /* id */Json_decode.field("id", Json_decode.$$int, json),
          /* name */Json_decode.field("name", Json_decode.string, json),
          /* age */Json_decode.optional(function (param) {
                return Json_decode.field("age", Json_decode.$$int, param);
              }, json),
          /* address */Json_decode.optional(function (param) {
                return Json_decode.field("address", parseAddress, param);
              }, json)
        ];
}

var data = "{\"id\":1,\"name\":\"brad\",\"age\":27,\"address\":{\"city\":\"city1\",\"state\":\"state1\"}}";

console.log(parsePerson(JSON.parse(data)));

exports.arrayOfInts         = arrayOfInts;
exports.mapJsonObjectString = mapJsonObjectString;
exports.sum                 = sum;
exports.parseAddress        = parseAddress;
exports.parsePerson         = parsePerson;
exports.data                = data;
/*  Not a pure module */
